src/mcp_atlassian/utils/oauth.py

```diff
@@ -0,0 +1,307 @@
+        """
+        if not self.refresh_token:
+            logger.error("No refresh token available")
+            return False
+
+        try:
+            payload = {
+                "grant_type": "refresh_token",
+                "client_id": self.client_id,
+                "client_secret": self.client_secret,
+                "refresh_token": self.refresh_token,
+            }
+
+            logger.debug("Refreshing access token...")
+            response = requests.post(TOKEN_URL, data=payload)
+            response.raise_for_status()
+
+            # Parse the response
+            token_data = response.json()
+            self.access_token = token_data["access_token"]
+            # Refresh token might also be rotated
+            if "refresh_token" in token_data:
+                self.refresh_token = token_data["refresh_token"]
+            self.expires_at = time.time() + token_data["expires_in"]
+
+            # Save the tokens
+            self._save_tokens()
+
+            return True
+        except Exception as e:
+            logger.error(f"Failed to refresh access token: {e}")
+            return False
+
+    def ensure_valid_token(self) -> bool:
+        """Ensure the access token is valid, refreshing if necessary.
+
+        Returns:
+            True if the token is valid (or was refreshed successfully), False otherwise.
+        """
+        if not self.is_token_expired:
+            return True
+        return self.refresh_access_token()
+
+    def _get_cloud_id(self) -> None:
+        """Get the cloud ID for the Atlassian instance.
+
+        This method queries the accessible resources endpoint to get the cloud ID.
+        The cloud ID is needed for API calls with OAuth.
+        """
+        if not self.access_token:
+            logger.debug("No access token available to get cloud ID")
+            return
+
+        try:
+            headers = {"Authorization": f"Bearer {self.access_token}"}
+            response = requests.get(CLOUD_ID_URL, headers=headers)
+            response.raise_for_status()
+
+            resources = response.json()
+            if resources and len(resources) > 0:
+                # Use the first cloud site (most users have only one)
+                # For users with multiple sites, they might need to specify which one to use
+                self.cloud_id = resources[0]["id"]
+                logger.debug(f"Found cloud ID: {self.cloud_id}")
+            else:
+                logger.warning("No Atlassian sites found in the response")
+        except Exception as e:
+            logger.error(f"Failed to get cloud ID: {e}")
+
+    def _save_tokens(self) -> None:
+        """Save the tokens to a file for later use.
+
+        This allows the tokens to be reused between runs without requiring
+        the user to go through the authorization flow again.
+        """
+        try:
+            # Create the directory if it doesn't exist
+            token_dir = Path.home() / ".mcp-atlassian"

```

Comment:
If you decide to use keyring for storage, the token loading function would need to be updated as well to retrieve from the secure storage instead of the JSON file.